### ISSUE TITLE
x/tools/gopls: bad performance with multiple workspace folders

### ISSUE THREAD

**Issue by bensussman** [golang/go#40295]
Describe what you observed.

GPLS uses 200% of my CPU (both cores) indefinitely whenever I rebuild my golang in my terminal. The only way to make my computer responsive again is to force-quit gopls. Importantly, the usability of gopls (go-to-definition in VSCode for example) is *broken* until I force-quit the process enough times such that the CPU goes down. Only then does it actually work.

Please attach the stack trace from the crash.
A window with the error message should have popped up in the lower half of your screen.
Please copy the stack trace from that window and paste it in this issue.

```
[Info  - 3:45:30 PM] 2020/07/17 15:45:30 go/packages.Load
	snapshot=0
	directory=/Users/ben/spell/src/spell
	query=[./... builtin]
	packages=191

[Info  - 3:45:36 PM] 2020/07/17 15:45:36 go/packages.Load
	snapshot=6
	directory=/Users/ben/spell
	query=[./]
	packages=1

[Info  - 3:45:37 PM] 2020/07/17 15:45:37 go/packages.Load
	snapshot=6
	directory=/Users/ben/spell
	query=[file=/Users/ben/spell/docker/runs/save/main.go file=/Users/ben/spell/src/spell/constants/paths/paths.go file=/Users/ben/spell/src/spell/db/run.go file=/Users/ben/spell/src/spell/machine/remote_operations.go file=/Users/ben/spell/src/spell/runorchestrator/worker.go file=/Users/ben/spell/src/spell/servers/api/resources/ls.go]
	packages=1

[Info  - 3:45:37 PM] 2020/07/17 15:45:37 go/packages.Load
	snapshot=6
	package_path="command-line-arguments"
	files=[/Users/ben/spell/docker/runs/save/main.go]

[Error - 3:45:37 PM] 2020/07/17 15:45:37 no dep handle: no metadata for gocloud.dev/blob
	package="gocloud.dev/blob"

[Error - 3:45:37 PM] 2020/07/17 15:45:37 no dep handle: no metadata for gocloud.dev/blob/s3blob
	package="gocloud.dev/blob/s3blob"

[Error - 3:45:37 PM] 2020/07/17 15:45:37 no dep handle: no metadata for gocloud.dev/blob/gcsblob
	package="gocloud.dev/blob/gcsblob"

[Error - 3:45:37 PM] 2020/07/17 15:45:37 no dep handle: no metadata for github.com/fluent/fluent-logger-golang/fluent
	package="github.com/fluent/fluent-logger-golang/fluent"

[Error - 3:45:37 PM] 2020/07/17 15:45:37 no dep handle: no metadata for github.com/aws/aws-sdk-go/aws/awserr
	package="github.com/aws/aws-sdk-go/aws/awserr"

[Error - 3:45:40 PM] Connection to server got closed. Server will not be restarted.
```

**Comment by bensussman** [comment]
Yes @stamblerre it does seem like closing out my multi-workspace-folder VSCode instance, and instead opening *just* the directory that contains my golang source code (in my case `~/spell/src/spell`, where my `go.mod` lives).

This is not a great solution for me, because it means that I have to have at least two windows open:

1. with Golang code dir
2. with everything else (and this has to be a multi-directory workspace because `python` and `js` are siblings).

I think an ideal fix would be to explicitly enable or disable golang/gopls within each directory. That way i could have my one-giant-workspace, but gopls would only be enabled for the one golang directory and not the parent.

**Comment by stamblerre** [comment]
Glad to hear it's working better. I agree it's not an ideal solution, and our plan is to address this shortcoming of `gopls` for the v1.0 release (https://github.com/golang/go/issues/32394). Our plan is to make sure that `gopls` works without needing to add your module as a workspace folder, so I think that will resolve this issue. Unfortunately, you may need to use this workaround for the time being.

**Comment by stamblerre** [comment]
Closing this, as it will be addressed with the more fundamental shifts in v1.0.
